walletnode/services/artworksearch/task.go

```diff
@@ -0,0 +1,212 @@
+
+		group.Go(func() error {
+
+			data, found, err := task.fetchThumbnail(ctx, res)
+			if err != nil {
+				log.WithContext(ctx).WithField("txid", res.TXID).WithError(err).Error("Fetch Thumbnail")
+
+				return fmt.Errorf("txid: %s - err: %s", res.TXID, err)
+			}
+
+			if found {
+				res.Thumbnail = data
+			} else {
+				log.WithContext(ctx).WithField("txid", res.TXID).Warn("Thumbnail not found")
+			}
+
+			// Post on result channel
+			task.resultChan <- res
+
+			log.WithContext(ctx).WithField("search_result", res).Debug("Posted search result")
+
+			return nil
+		})
+	}
+
+	return group.Wait()
+}
+
+// fetchThumbnail pulls artwork thumbnail & sets RegTicketSearch.Thumbnail if found
+func (task *Task) fetchThumbnail(ctx context.Context, res *RegTicketSearch) (data []byte, found bool, err error) {
+	hash := res.RegTicketData.ArtTicketData.AppTicketData.ThumbnailHash
+	key := b58.Encode(hash)
+
+	return task.p2pClient.Get(ctx, key)
+}
+
+// regTicket pull art registration ticket from cNode & decodes base64 encoded fields
+func (task *Task) regTicket(ctx context.Context, ticket *pastel.ActTicket) (pastel.RegTicket, error) {
+	regTicket, err := task.pastelClient.RegTicket(ctx, ticket.ActTicketData.RegTXID)
+	if err != nil {
+		return regTicket, fmt.Errorf("fetch: %s", err)
+	}
+
+	if err := fromBase64(string(regTicket.RegTicketData.ArtTicket),
+		&regTicket.RegTicketData.ArtTicketData); err != nil {
+		return regTicket, fmt.Errorf("convert art ticket: %s", err)
+	}
+
+	if err := fromBase64(string(regTicket.RegTicketData.ArtTicketData.AppTicket),
+		&regTicket.RegTicketData.ArtTicketData.AppTicketData); err != nil {
+		return regTicket, fmt.Errorf("convert app ticket: %s", err)
+	}
+
+	return regTicket, nil
+}
+
+// filterRegTicket filters ticket against request params & checks if its a match
+func (task *Task) filterRegTicket(regTicket pastel.RegTicket) (srch *RegTicketSearch, matched bool) {
+	if !inIntRange(regTicket.RegTicketData.ArtTicketData.AppTicketData.RarenessScore,
+		task.request.MinRarenessScore, task.request.MaxRarenessScore) {
+		return srch, false
+	}
+
+	if !inIntRange(regTicket.RegTicketData.ArtTicketData.AppTicketData.NSFWScore,
+		task.request.MinNsfwScore, task.request.MaxNsfwScore) {
+		return srch, false
+	}
+
+	if !inIntRange(regTicket.RegTicketData.ArtTicketData.AppTicketData.TotalCopies,
+		task.request.MinCopies, task.request.MaxCopies) {
+		return srch, false
+	}
+
+	regSearch := &RegTicketSearch{
+		RegTicket: &regTicket,
+	}
+
+	return regSearch.Search(task.request)
+}
+
+// addMatchedResult adds to search result
+func (task *Task) addMatchedResult(res *RegTicketSearch) {
+	task.searchResMutex.Lock()
+	defer task.searchResMutex.Unlock()
+
+	task.searchResult = append(task.searchResult, res)
+}
+
+// Error returns task err
+func (task *Task) Error() error {
+	return task.err
+}
+
+// SubscribeSearchResult returns a new search resultof the state.
+func (task *Task) SubscribeSearchResult() <-chan *RegTicketSearch {
+
```

Comment:
a little bit confusing here.
The Subscribe name raise a thought that allow multiple subscribers can listen the change.
But in this case we allow only 1 subscriber.
Anw. no big problem now.